hibernate-core/src/main/java/org/hibernate/type/descriptor/jdbc/spi/DescriptiveJsonGeneratingVisitor.java

```diff
@@ -0,0 +1,216 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ * Copyright Red Hat Inc. and Hibernate Authors
+ */
+package org.hibernate.type.descriptor.jdbc.spi;
+
+import org.hibernate.bytecode.enhance.spi.LazyPropertyInitializer;
+import org.hibernate.collection.spi.CollectionSemantics;
+import org.hibernate.collection.spi.PersistentCollection;
+import org.hibernate.collection.spi.PersistentMap;
+import org.hibernate.internal.util.collections.IdentitySet;
+import org.hibernate.metamodel.mapping.CollectionPart;
+import org.hibernate.metamodel.mapping.EntityIdentifierMapping;
+import org.hibernate.metamodel.mapping.EntityMappingType;
+import org.hibernate.metamodel.mapping.EntityValuedModelPart;
+import org.hibernate.metamodel.mapping.PluralAttributeMapping;
+import org.hibernate.metamodel.mapping.SelectableMapping;
+import org.hibernate.metamodel.mapping.ValuedModelPart;
+import org.hibernate.metamodel.mapping.internal.BasicValuedCollectionPart;
+import org.hibernate.metamodel.mapping.internal.EmbeddedAttributeMapping;
+import org.hibernate.persister.collection.CollectionPersister;
+import org.hibernate.type.descriptor.WrapperOptions;
+import org.hibernate.type.descriptor.jdbc.JsonHelper;
+import org.hibernate.type.format.JsonDocumentWriter;
+
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.function.Consumer;
+
+import static org.hibernate.Hibernate.isInitialized;
+
+/**
+ * Extension of {@link JsonHelper} that can be used to generate
+ * descriptive JSON output, that is more appropriate for representing
+ * complex data structures for displaying purposes.
+ */
+public class DescriptiveJsonGeneratingVisitor extends JsonGeneratingVisitor {
+
+	private Map<String, IdentitySet<Object>> circularityTracker;
```

Comment:
Isn't the "main" entry point to `DescriptiveJsonGeneratingVisitor` always going to be `serializeEntity`? If so, there is no need to make the initialization lazy.

Also, it would be nice to have a test that verifies the expectation of this `DescriptiveJsonGeneratingVisitor` to avoid breaking it.

```suggestion
	private final Map<String, IdentitySet<Object>> circularityTracker = new HashMap<>();
```

Reply:
In that case, forget my suggestion about eager initialization of the map. The tests would be nice though.

Reply:
Added a simple test.